log-warper/src/Log/Configuration/Extension.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Log.Configuration.Extension
+       ( Extension (..)
+       , ElemExtension
+       , ExtensionType
+
+       , HasRotation
+       , HasTerminal
+       ) where
+
+import Universum
+
+import Data.Type.Bool (If)
+
+-- | Kind for extension points which are used in logger configuration.
+data Extension = Rotation | Terminal
+
+-- | Type family for checking whether given 'Extension' is in list of extensions.
+type family ElemExtension (ext :: Extension) (exts :: [Extension]) :: Bool where
```

Comment:
I don't 100% understand what those extensions are, but what is the motivation to track them on the type-level?

Reply:
@int-index Those extensions solve the following problem:

In `log-warper` we want to have single `.yaml` file which configures every part of logging. And with monoidal `LogAction` you need to perform two steps:

1. Specify options in `.yaml` file (for example, `rotation` options).
2. Don't forget to add `<> logRotation` to action you want to run.

So `Configuration` should be flexible in terms that it needs to support both presence of rotation paramenters and absence. Because I don't want this behavior: when you specify `rotation` parameters in `.yaml` file and don't see logger rotation. The idea is that when you add `<> logRotation` to resulting action, the type of `exts` type variable for `Configuration exts` should be infered as

```haskell
HasRotation exts => Configuration exts
```

And if you don't add `<> logRotation` but you specify `rotation` parameters in `.yaml` file you will see runtime parsing error. So, basically, this `Extension` mechanism allows you to extend `Configuration` data type by required parts of parameters and prevents your from accidentally forget to do one of both things: (1) specify parameters in `.yaml` or (2) add `<> logRotation`.

I'm not claiming that this solution is the best one. For example, we can remove this type-level magic and just pass `RotationParameters` explicitly to `logRotation`. But with this solution you now have some parts of configuration through `yaml` and some parts through code or multiple `yaml` files and I don't like any of alternative solutions.

Another solution: just as before. Store only `Maybe SomeParameters` in `Configuration`. And configure logging through boolean flags. But I don't really like this solution because it's too fragile and can be broken at any time (as you noticed earlier). I really like to replace:

```haskell
let myAction = logFile
```

with

```haskell
let myAction = logFile <> logStdout
```

to get stdout output instead of changing value of `.yaml` configuration.

Reply:
I still don't understand why we have to track the extension at compile-time if the error will be triggered at runtime.

Reply:
I'm trying to understand your example but there isn't `logRotation` defined anywhere in this PR (or GitHub hides it).

But I think I understand the basic problem you're dealing with. You want to compose loggers `l1 <> l2 <> l3` etc, but each can have its own section in the configuration YAML file, correct? Well, in this case you definitely can't afford to have a closed kind `data Extension = ...`, what if the user wants to define his own extension?

Reply:
@int-index I want the following behavior:

1. Parse **_successfully_** rotation parameters if they **are specified** and I **need** them.
2. **_Fail_** parsing when rotation parameters **are specified** and when I **don't need** them.
3. **_Fail_** on parsing when rotation parameters **aren't specified** but I **need** them.
4. Parse **_successfully_** when rotation parameters **aren't specified** and I **don't need** them.

There's supposed to be the only way with current design to tell that you **need** rotation parameters: `<> logRotation`. I don't know better way to require rotation parameters from configuration when you need to manually add additional action.

With previous design rotation was enabled only by `.yaml`. If you have rotation parameters you will have logging rotation. If you don't have rotation parameters you won't have logger rotation. The same design can be used with new interface as well if you like to remove all this type-level stuff. But instead of having on user side:

```haskell
let myAction = logFile <> logRotation
```

You will always have on user side the following:

```haskell
let myAction = logConfigurationAction
```

and `<> logRotation` will be performed internally in `log-warper` library, depending on whether you specified rotation parameters in `.yaml` file or not.

Just a matter of trade-off which design people like more and which behavior they want more and is it worth to have type-level magic.

Reply:
>   Parse successfully rotation parameters if they are specified and I need them.
    Fail parsing when rotation parameters are specified and when I don't need them.
    Fail on parsing when rotation parameters aren't specified but I need them.
    Parse successfully when rotation parameters aren't specified and I don't need them.

Yes, I get this. What I don't get is why you need compile-time machinery for this check. I don't know what type `logRotation` in your example has, but it can be something like this:

```
(message -> m (), Set Extension)
```

then you'll have

```
logFile = (\_ -> writeFile ..., Set.empty)
logRotation = (\_ -> ..., Set.singleton Rotation)
```

When you combine these monoidally, the result will contain a `Set` with all _required_ configuration sections.

At the same time, after you parse a YAML file, you can easily get a `Set` with all _provided_ configuration sections.

Then you just compare these sets and throw an error if they don't match.

I'm not telling you to use this solution with `Set Extension` specifically, I'm just saying that I see no reason to make types more complicated for a check that is performed at runtime.

Reply:
> In my defense 

It's not _your_ defense, we're just discussing code, relax :slightly_smiling_face:

Reply:
>  Fail parsing when rotation parameters are specified and when I don't need them.

With the current approach, I don't really understand how you're going to achieve this. Okay, so doing `<> logRotation` will add a `HasRotation exts` constraint.

But in case I *don't* do `<> logRotation`, what will create a `HasNoRotation exts` constraint? Where will the `exts` list be specified in the end? That's something I don't understand.

Reply:
> configuration-is-monoid approach

In new `log-warper` configuration is no longer a monoid.

> Can we make a datatype per extension and a special typeclass which determines the extra action to be mappended?

This is basically an _extensible records_ problem. It has various solutions. Maybe something like that can be implemented but I didn't put much thought into this.

Reply:
>This is basically an extensible records problem. 

Not really. This may sound too ambitious, but I think that there is no need to unite all "extensions" into one record. I can imagine the existential wrapper `SomeConf = forall x. HasNeededActionFunctionality x => SomeConf x` over some extension config (`RotationConf` or `StdoutConf`). We can read these from the file directly and then pass it to logger initializer. It may sound (and be) a complete bullshit, but does it at least seem to be a direction to think about? IIRC something similar was done in the old log-warper with handlers -- they were abstract, user-defined, abstracted into existentially qualified `HandlerT`.